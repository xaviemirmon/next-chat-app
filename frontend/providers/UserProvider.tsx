"use client";

import { createContext, useState, useContext, ReactNode } from "react";

// Define the shape of the User Context
interface UserContextType {
  user: number | null;
  updateUser: (newUserData: React.SetStateAction<number | null>) => void;
  logoutUser: () => void;
}

// Create the context with the defined type
const UserContext = createContext<UserContextType | null>(null);

// Custom hook to access the User Context
export const useUser = () => {
  const context = useContext(UserContext);
  if (!context) {
    throw new Error("useUser must be used within a UserProvider");
  }
  return context;
};

export const UserProvider = ({ children }: { children: ReactNode }) => {
  // Define the state that holds the current user data
  const [user, setUser] = useState<number | null>(null);

  // Function to update user (e.g., after login)
  const updateUser = (newUserData: React.SetStateAction<number | null>) => {
    setUser(newUserData);
  };

  // Function to log out user
  const logoutUser = () => {
    setUser(null);
  };

  return (
    <UserContext.Provider value={{ user, updateUser, logoutUser }}>
      {children}
    </UserContext.Provider>
  );
};
