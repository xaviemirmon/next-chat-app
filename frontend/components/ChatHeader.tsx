import Link from "next/link";
import styles from "./ChatHeader.module.css";
import { Button } from "@/components/Button";
import { IoChevronBack, IoEllipsisHorizontal } from "react-icons/io5";
import Image from "next/image";
import { Dispatch, SetStateAction } from "react";
import { AppRouterInstance } from "next/dist/shared/lib/app-router-context.shared-runtime";
import { useUser } from "@/providers/UserProvider";

export const ChatHeader = ({
  userName,
  setLoading,
  router,
}: {
  userName: string | undefined;
  userId: number | undefined;
  setLoading: Dispatch<SetStateAction<boolean>>;
  router: AppRouterInstance;
}) => {
  const { user, logoutUser } = useUser();
  return (
    <div className={styles.header}>
      <div className={styles.profileLinks}>
        <Button
          variant="icon"
          onClick={() => {
            setLoading(true);
            logoutUser();
            router.push("/dashboard");
          }}
        >
          <IoChevronBack color={"#aeaeae"} />
        </Button>
        <div className={styles.userInfo}>
          <Image
            src={`https://i.pravatar.cc/30?img=${user}`}
            alt="Profile pic"
            height={30}
            width={30}
          />
          {userName}
        </div>
        <Button variant="icon">
          <IoEllipsisHorizontal color={"#aeaeae"} />
        </Button>
      </div>

      <Tabs />
    </div>
  );
};

const Tabs = () => (
  <div className={styles.tabs}>
    <Link href={"#"} className={`${styles.active} ${styles.tab}`}>
      Chat
    </Link>
    <Link href={"#"} className={`${styles.tab}`}>
      Profile
    </Link>
  </div>
);
