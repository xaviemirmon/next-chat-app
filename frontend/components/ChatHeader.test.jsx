import { render } from "@testing-library/react";
import { ChatHeader } from "./ChatHeader";
import { useUser } from "@/providers/UserProvider";

// Mocking the useUser hook
jest.mock("@/providers/UserProvider", () => ({
  useUser: jest.fn(),
}));

describe("ChatHeader component", () => {
  const mockSetLoading = jest.fn();
  const mockRouter = { push: jest.fn() };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Matches snapshot when user is logged in", () => {
    useUser.mockReturnValue({
      user: 1,
      logoutUser: jest.fn(),
    });

    const { asFragment } = render(
      <ChatHeader
        userName="Test User"
        setLoading={mockSetLoading}
        router={mockRouter}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Matches snapshot when userName is undefined", () => {
    useUser.mockReturnValue({
      user: 1,
      logoutUser: jest.fn(),
    });

    const { asFragment } = render(
      <ChatHeader
        userName={undefined}
        setLoading={mockSetLoading}
        router={mockRouter}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Matches snapshot when user is not logged in", () => {
    useUser.mockReturnValue({
      user: null,
      logoutUser: jest.fn(),
    });

    const { asFragment } = render(
      <ChatHeader
        userName="Test User"
        setLoading={mockSetLoading}
        router={mockRouter}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
