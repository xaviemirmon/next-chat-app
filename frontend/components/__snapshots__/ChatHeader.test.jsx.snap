// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatHeader component Matches snapshot when user is logged in 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <div
      class="profileLinks"
    >
      <button
        class="icon"
      >
        <svg
          color="#aeaeae"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(174, 174, 174);"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M328 112 184 256l144 144"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="48"
          />
        </svg>
      </button>
      <div
        class="userInfo"
      >
        <img
          alt="Profile pic"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3D1&w=64&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3D1&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3D1&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
        Test User
      </div>
      <button
        class="icon"
      >
        <svg
          color="#aeaeae"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(174, 174, 174);"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="256"
            cy="256"
            r="48"
          />
          <circle
            cx="416"
            cy="256"
            r="48"
          />
          <circle
            cx="96"
            cy="256"
            r="48"
          />
        </svg>
      </button>
    </div>
    <div
      class="tabs"
    >
      <a
        class="active tab"
        href="#"
      >
        Chat
      </a>
      <a
        class="tab"
        href="#"
      >
        Profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChatHeader component Matches snapshot when user is not logged in 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <div
      class="profileLinks"
    >
      <button
        class="icon"
      >
        <svg
          color="#aeaeae"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(174, 174, 174);"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M328 112 184 256l144 144"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="48"
          />
        </svg>
      </button>
      <div
        class="userInfo"
      >
        <img
          alt="Profile pic"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3Dnull&w=64&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3Dnull&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3Dnull&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
        Test User
      </div>
      <button
        class="icon"
      >
        <svg
          color="#aeaeae"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(174, 174, 174);"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="256"
            cy="256"
            r="48"
          />
          <circle
            cx="416"
            cy="256"
            r="48"
          />
          <circle
            cx="96"
            cy="256"
            r="48"
          />
        </svg>
      </button>
    </div>
    <div
      class="tabs"
    >
      <a
        class="active tab"
        href="#"
      >
        Chat
      </a>
      <a
        class="tab"
        href="#"
      >
        Profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChatHeader component Matches snapshot when userName is undefined 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <div
      class="profileLinks"
    >
      <button
        class="icon"
      >
        <svg
          color="#aeaeae"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(174, 174, 174);"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M328 112 184 256l144 144"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="48"
          />
        </svg>
      </button>
      <div
        class="userInfo"
      >
        <img
          alt="Profile pic"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3D1&w=64&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3D1&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fi.pravatar.cc%2F30%3Fimg%3D1&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
      </div>
      <button
        class="icon"
      >
        <svg
          color="#aeaeae"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(174, 174, 174);"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="256"
            cy="256"
            r="48"
          />
          <circle
            cx="416"
            cy="256"
            r="48"
          />
          <circle
            cx="96"
            cy="256"
            r="48"
          />
        </svg>
      </button>
    </div>
    <div
      class="tabs"
    >
      <a
        class="active tab"
        href="#"
      >
        Chat
      </a>
      <a
        class="tab"
        href="#"
      >
        Profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;
